initial commit for event bus doc

Hamstrings opened the pull request:
A light version of event docs. Also it requires refactoring. Just html format of UI, no styling, just skeleton to get stuff collaborating.

Hamstrings committed:
initial commit

Burgyn commented on tests/Kros.EvenBusDoc.Test.csproj:
```diff
@@ -0,0 +1,31 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Rename project `Kros.EventBusDoc.Test` to `Kros.EventBusDoc.Generator.Tests`

Burgyn commented on tests/Kros.EvenBusDoc.Test.csproj:
```diff
@@ -0,0 +1,31 @@
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <DocumentationFile>C:\Develop\bus\tests\resources\XML\Kros.EvenBusDoc.Test.xml</DocumentationFile>
```
Add suppress warnings 1701;1702;1591. 1591 - suppress XML missing commentar warning in test project.

```
  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
    <DocumentationFile>C:\Develop\bus\tests\resources\XML\Kros.EvenBusDoc.Test.xml</DocumentationFile>
    <NoWarn>1701;1702;1591</NoWarn>
  </PropertyGroup>
```

Burgyn commented on tests/EventBusDocTypeResolverShould.cs:
```diff
@@ -0,0 +1,207 @@
+        private IXmlReader _xmlReader;
+        private IBusentProvider _busetnProvider;
+
+        private const string XML_PATH = @"C:\Develop\bus\tests\resources\XML\Kros.EvenBusDoc.Test.xml";
```
Remove `XML_PATH` constant and load resource from assembly.

Burgyn commented on src/Kros.EventBusDoc.Generator/Middleware/Extensions/KrosEventBusDocBuilderExtensions.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace Kros.EventBusDoc.Generator.Middleware.Extensions
+{
+    public static class KrosEventBusDocBuilderExtensions
```
Premenoval by som tú triedu štandardne `ApplicationBuilderExtensions`

Burgyn commented on src/Kros.EventBusDoc.Generator/Middleware/Extensions/EventBusDocGenSerCollExtensions.cs:
```diff
@@ -0,0 +1,44 @@
+            return services;
+        }
+
+        public static void ConfigureEventBusDocGen(
```
Táto metóda nie je potrebná.

Burgyn commented on src/Kros.EventBusDoc.Generator/Middleware/Extensions/EventBusDocGenSerCollExtensions.cs:
```diff
@@ -0,0 +1,44 @@
+           Action<EventBusDocGenOptions> setupAction = null)
+        {
+            // Add Mvc convention to ensure ApiExplorer is enabled for all actions
+            services.Configure<MvcOptions>(c =>
```
Myslím si, že to nepotrebujeme. Ak to naozaj nepotrebujeme tak to odstráňme a aj tú triedy, ktorú k tomu vytvárame.

Burgyn commented on src/Kros.EventBusDoc.Generator/Middleware/Extensions/EventBusDocGenSerCollExtensions.cs:
```diff
@@ -0,0 +1,44 @@
+            services.Configure<MvcOptions>(c =>
+                c.Conventions.Add(new EventBusDocAppConvention()));
+
+            // Register generator and it's dependencies
```
Takéto komentáre sa snažíme minimalizovať. často časom prestanú byť aktuálne. Ak je to možné, radšej daný kus kódu zabaliť do metódy, ktorú výstižne pomenujeme.

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentAnnotation/EventBusConsumeerAttribute.cs:
```diff
@@ -0,0 +1,13 @@
+namespace Kros.EventBusDoc.Generator.BusentAnnotation
+{
+    [AttributeUsage(AttributeTargets.Assembly)]
+    public class EventBusConsumeerAttribute : EventBusBaseAttribute
```
Dve ee v slove consumeer :-)

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentAnnotation/EventBusCommandSenderAttribute.cs:
```diff
@@ -0,0 +1,14 @@
+namespace Kros.EventBusDoc.Generator.BusentAnnotation
+{
+    [AttributeUsage(AttributeTargets.Assembly)]
+    public class EventBusCommandSenderAttribute : EventBusBaseAttribute
```
Asi by sme si mali sadnúť aj s @satano a dohodnúť si názvy. 
Takto mi to nesedí, ale neviem či môj návrh je pelší :-(
Môj návrh:
```
[assembly: EventBusCommandPublish(EventType = typeof(ICommandOne))]
[assembly: EventBusCommandSubscribe(EventType = typeof(ICommandTwo))]
[assembly: EventBusMessagePublish(EventType = typeof(IMessageOne))]
[assembly: EventBusMessageSubscribe(EventType = typeof(IMessageTwo))]
```

Burgyn commented on tests/EventBusDocTypeResolverShould.cs:
```diff
@@ -0,0 +1,207 @@
+
+        #endregion //end of Initialization for tests
+
+        #region Tests For Collections Events, Commands, ResolvedTypes
```
Testy z tohto regiónu by mali byť vo vlastnej triede, pretože pravdepodobne testujú niečo iné.

Burgyn commented on tests/EventBusDocTypeResolverShould.cs:
```diff
@@ -0,0 +1,207 @@
+
+        #endregion // end fo Tests For Collections Events, Commands, ResolvedTypes
+
+        #region Inline Tests For Reading Summary & Remarks From Xml
```
Testy v tomto regióne testujú `BXmlReader` tak majú byť v samostatnej triede.

Burgyn commented on tests/EventBusDocTypeResolverShould.cs:
```diff
@@ -0,0 +1,207 @@
+            var result = _xmlReader.GetSummary(classType);
+
+            result.Equals(expected).Should().Be(testResult);
+
```
Prázdny riadok na konci metódy nepoužívame. Je to na veľa miestach.

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/XmlReader/BXmlReader.cs:
```diff
@@ -0,0 +1,74 @@
+        public void Load(string documentPath)
+        {
+            Document = new XmlDocument();
+            try
```
`Try` tu nie je potrebný.

Ak by aj bol potrebný tak exception sa nepreposiela `throw e`, ale len `throw`. Ak chceš vysvetlím rozdiel.

Hamstrings committed:
the little typo chenge applying decorator for html raw data ....

Burgyn commented on tests/EventBusDocTypeResolverShould.cs:
```diff
@@ -0,0 +1,207 @@
+
+namespace Kros.EvenBusDoc.Test
+{
+    public class EventBusDocTypeResolverShould
```
Chýba mi tu taký jasný test, kde nasimulujem assembly s atribútmi, zavolám danú metódu a dostanem json. Následne ten json porovnám s nejakým očakávaným.

Burgyn commented on src/Kros.EventBusDoc.Generator/Middleware/EventBusDocOptions.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public class EventBusDocOptions
+    {
+        public string RouteTemplate { get; set; } = "busent/{documentName}/busent.json";
```
Naozaj `{documentName}`, alebo radšej `{version}`?

Burgyn commented on src/Kros.EventBusDoc.Generator/Middleware/EventBusDocGeneratorMiddleware.cs:
```diff
@@ -0,0 +1,101 @@
+    public class EventBusDocGeneratorMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly ILogger<EventBusDocGeneratorMiddleware> _logger;
```
V tejto triede je veľa nepoužívaných premenných.

Burgyn commented on src/Kros.EventBusDoc.Generator/Middleware/EventBusDocGeneratorMiddleware.cs:
```diff
@@ -0,0 +1,101 @@
+            {
+                await RespondWithBusentJson(context.Response, provider);
+            }
+            catch (UnknownBusentDocument)
```
Táto výnimka sa reálne nikde nevyhadzuje, takže ju nemusíme odchytávať. Taktiež môžme zmazať danú triedu.

Burgyn commented on src/Kros.EventBusDoc.Generator/Middleware/EventBusDocGeneratorMiddleware.cs:
```diff
@@ -0,0 +1,101 @@
+
+        }
+
+        private bool RequestingBusentDocument(HttpRequest request, out string documentName)
```
Naozaj `documentName`? Nie náhodou `version`?

Burgyn commented on src/Kros.EventBusDoc.Generator/Middleware/Extensions/KrosEventBusDocBuilderExtensions.cs:
```diff
@@ -0,0 +1,38 @@
+using System.Threading.Tasks;
+using Kros.EventBusDoc.Generator.BusentScour.Interfaces;
+
+namespace Kros.EventBusDoc.Generator.Middleware.Extensions
```
Častokrát sa robí to, že táto extension sa dáva do namespace `Microsoft.AspNetCore.Builder`. Aby človek v `Startup` triede nemusel pridávať ďalšie namespaces.

Burgyn commented on src/Kros.EventBusDoc.Generator/Middleware/Extensions/EventBusDocGenSerCollExtensions.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace Kros.EventBusDoc.Generator.Middleware.Extensions
+{
+    public static class EventBusDocGenSerCollExtensions
```
Túto triedy by som nazval štandartne `ServiceCollectionExtensions` a namespace premenoval na `Microsoft.Extensions.DependencyInjection`.

Burgyn commented on src/Kros.EventBusDoc.Generator/Helpers/Extensions.cs:
```diff
@@ -0,0 +1,101 @@
+
+namespace Kros.EventBusDoc.Generator.Helpers
+{
+    public static class Extensions
```
Nazval by som to `TypeExtensions`. Niektoré metódy sa nepoužívajú. To čo sa nepoužíva tak preč s tým.

Burgyn replied:
podľa mňa môže byť táto trieda `internal`.

Burgyn commented on src/Kros.EventBusDoc.Generator/Helpers/Utils.cs:
```diff
@@ -0,0 +1,28 @@
+    /// <summary>
+    /// Generic util for enumeration through an Enum, should be faster then explicit castig and unboxig/boxing.
+    /// </summary>
+    public static class Utils
```
Je tu jediná metóda a tá sa nepoužíva, takže kľudne preč s ňou.

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/XmlReader/XmlReadExtensions.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace Kros.EventBusDoc.Generator.BusentScour.XmlReader
+{
+    public static class XmlReadExtensions
```
Podľa mňa môže byť táto trieda `internal`.

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/XmlReader/IXmlReader.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Kros.EventBusDoc.Generator.BusentScour.XmlReader
+{
+    public interface IXmlReader
```
Nebolo by lepšie to nazvať `IXmlDocumentationReader`?
Taktiež si myslím, že tieto veci môžu byť `internal`.

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/XmlReader/EXmlTypeTag.cs:
```diff
@@ -0,0 +1,30 @@
+namespace Kros.EventBusDoc.Generator.BusentScour.XmlReader
+{
+    [Flags]
+    public enum EXmlTypeTag
```
Nazval by som to `XmlDocumentationType` a dal by som to `internal`.

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/XmlReader/EXmlTypeTag.cs:
```diff
@@ -0,0 +1,30 @@
+        /// <summary>
+        /// marks a property in xml document
+        /// </summary>
+        P = 1,
```
Nepoužívajme takéto skratky. Dajme tu normálne `Property`, `Class`

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/XmlReader/BXmlReader.cs:
```diff
@@ -0,0 +1,74 @@
+
+namespace Kros.EventBusDoc.Generator.BusentScour.XmlReader
+{
+    public class BXmlReader : IXmlReader
```
Nazval by som to `XmlDocumentationReader` a dal by som to ako `internal`.

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/XmlReader/BXmlReader.cs:
```diff
@@ -0,0 +1,74 @@
+
+        #region Member Functions
+
+        public string GetSummary(string item)
```
Nepoužíva sa, preč s ňou.

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/XmlReader/BXmlReader.cs:
```diff
@@ -0,0 +1,74 @@
+            }
+        }
+
+        public string GetSummary(Type type, PropertyInfo property = null)
```
Používame Expression-bodied members:
```suggestion
        public string GetSummary(Type type, PropertyInfo property = null)
          => GetInnerElementString(type, EXmlTypeTag.Summary, property);
```

Aj na iných miestach kde sa to nachádza.

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/Scourers/ScourerBase.cs:
```diff
@@ -0,0 +1,159 @@
+namespace Kros.EventBusDoc.Generator.BusentScour.Scourers
+{
+
+    public abstract class ScourerBase : IScourer
```
`internal`?

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/Scourers/ScourerBase.cs:
```diff
@@ -0,0 +1,159 @@
+
+        private bool OmitThisType(Type type)
+        {
+            return (type.IsPrimitive || IsRegistered(type) || (type.BaseType is null && !type.IsInterface) || (type == typeof(string))) ? true : false;
```
V konvenciach máme, že máme maximálnu dĺžku riadku 130 znakov.

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/Scourers/AttributeFetcher.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace Kros.EventBusDoc.Generator.BusentScour.Scourers
+{
+    public class AttributeFetcher : IAttributeFetcher
```
Niektoré metódy tejto triedy sa nikde nepoužívajú.

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/Scourers/AttributeFetcher.cs:
```diff
@@ -0,0 +1,40 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Kros.EventBusDoc.Generator.BusentScour.Scourers
```
Všetky triedy v tomto namespace by mohli byť `internal`.

Hamstrings committed:
closed format of taggind html,
preparing the jquery extended function to style the main format of html

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/Generators/DeclarativeGenerator.cs:
```diff
@@ -0,0 +1,167 @@
+        {
+            get
+            {
+                _document = GenerateFormat("natvrdo");
```
??

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/Generators/DeclarativeGenerator.cs:
```diff
@@ -0,0 +1,167 @@
+        }
+
+
+        public OpenApiDocument GetBusent(string docName, string host = null, string basePath = null)
```
Nie je nikde použitá.

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/Generators/DeclarativeGenerator.cs:
```diff
@@ -0,0 +1,167 @@
+                Commands = from type in _scourer.Commands select type.Name,
+                Consumes = from type in _scourer.Consumes select type.Name,
+
+                Version = "v0",
```
Je možné to z vonku nastaviť?

Burgyn commented on src/Kros.EventBusDoc.Generator/BusentScour/Generators/DeclarativeGenerator.cs:
```diff
@@ -0,0 +1,167 @@
+
+            foreach (var prop in item.GetProperties())
+            {
+                descObj.Properties = descObj.Properties.Append(
```
nestačí tu zavolať nasledovné:
```
descObj.Properties.Append(
                        new Property()
                        {
                            Description = _xmlReader.GetSummary(item, prop),
                            Name = prop.Name,
                            IsPrimitive = prop.PropertyType.IsPrimitive,
                            Type = prop.PropertyType.GetName()
                        }
                    );
```

bez toho priradenia?

Burgyn reviewed (changes requested):
- [ ] Fix warning:
> Warning	MSB3270	There was a mismatch between the processor architecture of the project being built "MSIL" and the processor architecture of the reference

- [ ] Remove unused usings in files
- [ ] Turn on documentation generation in projects.
- [ ] Okomentovať jednotlivé triedy. (Všetky public triedy a metódy v hlavných projektoch musia byť okomentované)
- [ ] Pre json je typickejšie, že je camelcase
- [ ] Demo teraz má 3 eventy, ale v sekcií definition sa zobrazujú len dve.
- [ ] Testy na mojom PC neprechádzajú.

Hamstrings committed:
little research, and trying typing the extension for jQquery

Hamstrings committed:
added ui card style, next referencing types and navigation

Hamstrings committed:
the navigation for the definition list works, but not as required, needs to distinquish the event for clicking to expand and the scroll for the definition, therefore it needs to add a span with the href for the definition

Hamstrings committed:
added full name to the property structure, due to getting the proper navigation, if the two there are two types but from diferent assembly

Hamstrings committed:
a bad format occured when array of generic class as input, the formater returns the crazy string repre => problem then in referencing in the html

Hamstrings committed:
little workaround with replacing/escaping special chars, for definition (not detached) navigation according the types works ....

Hamstrings committed:
navigation from types -> types works ... a little problem no navigation if collapsed elements

Hamstrings committed:
navigation from definition to collapsed types definition works as expected

Hamstrings committed:
it has a header style.. no more styles yet

Hamstrings committed:
styled properties using the groupbox stylus to looks better

Hamstrings committed:
clean up the code such as spacing, lining, unused using statements etc.

Hamstrings committed:
SOLVING ISSUES:
TURNED ALL DOCUMENT GENERATION, REMOVED UNUSED USINGS IN FILES, JSON DEFAULT SETTINGS, FIX WARNING, WARNINGS FIX, REMOVED CONST XML_PATH, USED RESOURCES

Hamstrings committed:
RESOLVE ISSUE:
 RENAMED THE KROS.EVENBUSDOC.TEST -> KROS.EVENBUSDOC.GENERATOR.TEST, RENAMED KROSEVENTBUSDOCBUILDEREXTENSIONS -> ApplicationBuilderExtensions, REMOVED EventBusDocAppConvention,  EventBusConsumerAttribute RENAMED, TESTS DEVIDED INTO TO CLASSES, REMOVED UNNECCESSARY EMPTY LINES, REMOVED TRY IN LOAD, REMOVED NOT USED EXCEPTION

Hamstrings committed:
all possible changes done, the rest of renaming, still needs some visul refactor etc

Hamstrings committed:
service header column style

satano commented:
Tento PR rušíme a bude sa pozerať [UI Resolution](https://github.com/Kros-sk/Kros.BusEventDoc/pull/5).
